sycl/plugins/unified_runtime/ur/ur.hpp

```diff
@@ -76,5 +79,43 @@ class pi_mutex {
   }
 };
 
+/// SpinLock is a synchronization primitive, that uses atomic variable and
+/// causes thread trying acquire lock wait in loop while repeatedly check if
+/// the lock is available.
+///
+/// One important feature of this implementation is that std::atomic<bool> can
+/// be zero-initialized. This allows SpinLock to have trivial constructor and
+/// destructor, which makes it possible to use it in global context (unlike
+/// std::mutex, that doesn't provide such guarantees).
+class SpinLock {
+public:
+  void lock() {
+    while (MLock.test_and_set(std::memory_order_acquire))
+      std::this_thread::yield();
+  }
+  void unlock() { MLock.clear(std::memory_order_release); }
+
+private:
+  std::atomic_flag MLock = ATOMIC_FLAG_INIT;
+};
+
+// Helper for one-liner validation
+#define PI_ASSERT(condition, error)                                            \
+  if (!(condition))                                                            \
+    return error;
+
 // TODO: populate with target agnostic handling of UR platforms
 struct _ur_platform {};
+
+// Controls tracing UR calls from within the UR itself.
+extern bool PrintTrace;
+
+// Apparatus for maintaining immutable cache of platforms.
+//
+// Note we only create a simple pointer variables such that C++ RT won't
+// deallocate them automatically at the end of the main program.
+// The heap memory allocated for these global variables reclaimed only at
+// explicit tear-down.
+extern std::vector<zer_platform_handle_t> *PiPlatformsCache;
+extern SpinLock *PiPlatformsCacheMutex;
+extern bool PiPlatformCachePopulated;

```

Comment:
Do I get it right that things like this will be eventually renamed to have prefix UR/Ur instead of Pi?

Reply:
right, just wanted to keep changes as minimal as possible